Markdown updates for 3.0.4

lornajane opened the pull request:
We need to handle the changes to anchor links and headings (see #3548  for context). We also have #3596 to add tooling, which I will work on after we've done this initial cleanup.

The commits tell their own story, but this is all very repeatable and hopefully transferrable between branches. Shout out to @handrews who got me started with the anchor/link rewriting script.

The process goes like this:

1. Run [prettier](https://github.com/prettier/prettier) for formatting
    -   `prettier --write --single-quote 3.0.4.md`

2. Run [markdownlint](https://github.com/DavidAnson/markdownlint-cli2) to fix whatever it can
    - `markdownlint --fix 3.0.4.md`
    - Configuration file `.markdownlint.yaml`:
      ```YAML
            MD007:
              indent: 2

            MD012: false # allow blank lines

            MD013:
              line_length: 800
              tables: false

            MD024: false # duplicate headings
            MD033: false # inline HTML
      ```
3.  Manually fix additional markdownlint problems
    - heading levels aren't continuous MD001
    - code fences need a language MD040
    - table has the wrong number of cells MD056
4.  Take out the table of contents and its comments, replace with a single `<!-- toc -->`
5.  Run a magical one-off script to update/fix/rewrite/remove all our anchors and internal links. Basic idea:
     - make all our anchor links kebab-case
     - update all other internal document links
     - remove the ones that are in headings
     - add a lookup/override to fix things that either were inflected
       differently or we were using an anchor link that didn't match the
       title
     - (script is in comment)
     - use markdownlint again to check that this all worked because it can check internal links
6.  Put the TOC in with all the new structure and links:
    - `markdown-toc --maxdepth 4 -i 3.0.4.md`

It's 850 lines of change, I don't know how we're going to review it, but take a look!

lornajane commented:
Super special python script:

```python
from sys import argv
from pathlib import Path
import re

# this script tries to inflect the old links, some are just missing and we need to use the title's version instead
updates = {}
updates["revision-history"] = "appendix-a-revision-history"
updates["data-type-conversion"] = "appendix-b-data-type-conversion"
updates["using-r-f-c6570-implementations"] = "appendix-c-using-rfc6570-implementations"
updates["serializing-headers-and-cookies"] = "appendix-d-serializing-headers-and-cookies"
updates["percent-encoding-and-form-media-types"] = "appendix-e-percent-encoding-and-form-media-types"
updates["document-structure"] = "openapi-description-structure"
updates["oas-object"] = "openapi-object"
updates["components-security-schemes"] = "security-scheme-object"
updates["schema-composition"] = "composition-and-inheritance-polymorphism"
updates["http-codes"] = "http-status-codes"
updates["oas-document"] = "openapi-description"
updates["rich-text"] = "rich-text-formatting"
updates["relative-references"] = "relative-references-in-urls"
updates["runtime-expression"] = "runtime-expressions"
updates["runtime-expression-examples"] = "examples"



def kebab_it(c):
    if c.lower() != c: 
        return f'-{c.lower()}'
    return c

if __name__ == '__main__':
    text = Path(argv[1]).read_text()

    names = {}
    removals = {}
    for match in re.finditer(r'\n(.*)<a name="([^"]*)"', text):
        name = match.group(2)
        names[name] = ''.join([kebab_it(c) for c in name])

        # was it a heading? file it for removal
        if len(match.group(1)) and match.group(1)[0] == "#":
            removals[name] = True

    for current, replacement in names.items():
        if replacement in updates:
            replacement = updates[replacement]
        text = text.replace(f'(#{current})', f'(#{replacement})')

        # only remove if removal is indicated, otherwise update
        if current in removals:
            text = text.replace(f'<a name="{current}"></a>', '')
        else:
            text = text.replace(f'<a name="{current}"></a>', f'<a name="{replacement}"></a>')
            

    print(text)

```

Run it like: `python kebab_it.py 3.0.4.md > temp.md` and then if `temp.md` looks good, copy it back over `3.0.4.md`

handrews commented on versions/3.0.4.md:
```diff
@@ -208,62 +245,62 @@ These limitations are expected to be addressed in a future release.
 Formats such as `"email"`, `"uuid"`, and so on, MAY be used even though they are not defined by this specification.
-The OpenAPI Initiative also hosts a [Format Registry](https://spec.openapis.org/registry/format/) for formats defined by OAS users and other specifications.  Support for any registered format is strictly OPTIONAL, and support for one registered format does not imply support for any others.
+The OpenAPI Initiative also hosts a [Format Registry](https://spec.openapis.org/registry/format/) for formats defined by OAS users and other specifications. Support for any registered format is strictly OPTIONAL, and support for one registered format does not imply support for any others.
 
 Types that are not accompanied by a `format` property follow the type definition in the JSON Schema. Tools that do not recognize a specific `format` MAY default back to the `type` alone, as if the `format` is not specified.
 
 The formats defined by the OAS are:
 
-[`type`](#dataTypes) | [`format`](#dataTypeFormat) | Comments
------- | -------- | --------
-`integer` | `int32` | signed 32 bits
-`integer` | `int64` | signed 64 bits (a.k.a long)
-`number` | `float` | |
-`number` | `double` | |
-`string` | `byte` | base64 encoded characters - [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-4)
-`string` | `binary` | any sequence of octets
-`string` | `date` | As defined by `full-date` - [RFC3339](https://www.rfc-editor.org/rfc/rfc3339#section-5.6)
-`string` | `date-time` | As defined by `date-time` - [RFC3339](https://www.rfc-editor.org/rfc/rfc3339#section-5.6)
-`string` | `password` | A hint to obscure the value.
+| [`type`](#data-types) | [`format`](#data-type-format) | Comments                                                                                  |
+| -------------------- | --------------------------- | ----------------------------------------------------------------------------------------- |
+| `integer`            | `int32`                     | signed 32 bits                                                                            |
+| `integer`            | `int64`                     | signed 64 bits (a.k.a long)                                                               |
+| `number`             | `float`                     |                                                                                           |
+| `number`             | `double`                    |                                                                                           |
+| `string`             | `byte`                      | base64 encoded characters - [RFC4648](https://www.rfc-editor.org/rfc/rfc4648#section-4)   |
+| `string`             | `binary`                    | any sequence of octets                                                                    |
+| `string`             | `date`                      | As defined by `full-date` - [RFC3339](https://www.rfc-editor.org/rfc/rfc3339#section-5.6) |
+| `string`             | `date-time`                 | As defined by `date-time` - [RFC3339](https://www.rfc-editor.org/rfc/rfc3339#section-5.6) |
+| `string`             | `password`                  | A hint to obscure the value.                                                              |
 
-#### <a name="binaryData"></a>Working With Binary Data
+#### Working With Binary Data
 
 Two formats, `binary` and `byte`, describe different ways to work with binary data:
 
-* `binary` is used where unencoded binary data is allowed, such as when sending a binary payload as an HTTP message body, or as part of a `multipart/*` payload that allows binary parts
-* `byte` is used where binary data is embedded in a text-only format such as `application/json` or `application/x-www-form-urlencoded`
+- `binary` is used where unencoded binary data is allowed, such as when sending a binary payload as an HTTP message body, or as part of a `multipart/*` payload that allows binary parts
+- `byte` is used where binary data is embedded in a text-only format such as `application/json` or `application/x-www-form-urlencoded`
```
I'm a bit puzzled by this list indicator change.  I feel like we have always consistently used `*` for this and the change introduces a lot of noise.  It's not a hill I want to die on, but I'd prefer `*` (both because it's what I'm used to scaning visually for when I edit, and to trim down this diff)

lornajane replied:
`-` has evolved as a standard in markdown tools, the linter actually checks that the lists are consistent but I can totally enforce `*` instead if that's our normal.

handrews replied:
>  `-` has evolved as a standard in markdown tools

UGH why????  I'm a bit reluctant to start introducing digressions.  I do think `*` makes lists much easier to read in the source, but maybe that's because I'm more used to it.  It's just... bullet points... `*` looks like a bullet point, `-` does not!

handrews commented on versions/3.0.4.md:
```diff
@@ -153,40 +190,40 @@ In order to preserve the ability to round-trip between YAML and JSON formats, YA
 
 It is RECOMMENDED that the entry OpenAPI document be named: `openapi.json` or `openapi.yaml`.
 
-#### <a name="structuralInteroperability"></a>Structural Interoperability
+#### Structural Interoperability
 
-When parsing an OAD, JSON or YAML objects are parsed into specific Objects (such as [Operation Objects](#operationObject), [Response Objects](#responseObject), [Reference Objects](#referenceObject), etc.) based on the parsing context.  Depending on how references are arranged, a given JSON or YAML object can be parsed in multiple different contexts:
+When parsing an OAD, JSON or YAML objects are parsed into specific Objects (such as [Operation Objects](#operation-object), [Response Objects](#response-object), [Reference Objects](#reference-object), etc.) based on the parsing context. Depending on how references are arranged, a given JSON or YAML object can be parsed in multiple different contexts:
 
-* As a complete OpenAPI Description document
-* As the Object type implied by its parent Object within the document
-* As a reference target, with the Object type matching the reference source's context
+- As a complete OpenAPI Description document
+- As the Object type implied by its parent Object within the document
+- As a reference target, with the Object type matching the reference source's context
 
-If the same JSON/YAML object is parsed multiple times and the respective contexts require it to be parsed as _different_ Object types, the resulting behavior is _implementation defined_, and MAY be treated as an error if detected.  An example would be referencing an empty Schema Object under `#/components/schemas` where a Path Item Object is expected, as an empty object is valid for both types.   ...
+If the same JSON/YAML object is parsed multiple times and the respective contexts require it to be parsed as _different_ Object types, the resulting behavior is _implementation defined_, and MAY be treated as an error if detected. An example would be referencing an empty Schema Object under `#/components/schemas` where a Path Item Object is expected, as an empty object is valid for both types. Fo ...
 
-#### <a name="resolvingImplicitConnections"></a>Resolving Implicit Connections
+#### Resolving Implicit Connections
 
 Several features of this specification require resolution of non-URI-based connections to some other part of the OpenAPI Description (OAD).
 
 These connections are unambiguously resolved in single-document OADs, but the resolution process in multi-document OADs is _implementation-defined_, within the constraints described in this section.
 In some cases, an unambiguous URI-based alternative is available, and OAD authors are RECOMMENDED to always use the alternative:
 
-Source | Target | Alternative
------- | ------ | -----------
-[Security Requirement Object](#securityRequirementObject) `{name}` |  [Security Scheme Object](#securitySchemeObject) name under the [Components Object](#componentsObject) | _n/a_
-[Discriminator Object](#discriminatorObject) `mapping` _(implicit, or explicit name syntax)_ | [Schema Object](#schemaObject) name under the Components Object | `mapping` _(explicit URI syntax)_
-[Operation Object](#operationObject) `tags` | [Tag Object](#tagObject) `name` (in the Components Object) | _n/a_
-[Link Object](#linkObject) `operationId` | [Path Item Object](#pathItemObject) `operationId` | `operationRef`
+| Source                                                                                       | Target                                                                                                | Alternative                       |
+| -------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------------------------- | --------------------------------- |
+| [Security Requirement Object](#security-requirement-object) `{name}`                           | [Security Scheme Object](#security-scheme-object) name under the [Components Object](#components-object) | _n/a_                             |
+| [Discriminator Object](#discriminator-object) `mapping` _(implicit, or explicit name syntax)_ | [Schema Object](#schema-object) name under the Components Object                                       | `mapping` _(explicit URI syntax)_ |
+| [Operation Object](#operation-object) `tags`                                                  | [Tag Object](#tag-object) `name` (in the Components Object)                                            | _n/a_                             |
+| [Link Object](#link-object) `operationId`                                                     | [Path Item Object](#path-item-object) `operationId`                                                     | `operationRef`                    |
 
```
I have mixed feelings about this sort of table re-formatting.  On the one hand, it appeals to my sense of aesthetics.  On the other, it's very hard to get people to maintain it correctly.  I've gotten rather used to not bothering with consistent widths most of the time.  But I'm sure I bothered with it at least once, so... idk.

For me, it boils down to: "If a new contributor messes up this table spacing, will be block their PR until they fix it, and is that the sort of contributor experience we want to create?"

handrews replied:
Although if this is being done by `prettier` or some other script that we can set up to run as a pre-commit hook, then I'm 100% fine with it (and even the list item change, although I really do find `*` _much_ easier to spot when scanning text, as it is rarely used for anything other than lists, while `-` is used for many, many things and has less visual weight as weel).

lornajane replied:
it is all automatic, a couple of tables had varying numbers of columns but otherwise this is all auto-fix and we can get pre-commit and/or CI to do it or at least check it

handrews reviewed (commented):
A few quick first impressions...

mikekistler commented:
Do we need to keep the TOC? GitHub creates one for us ...

<img width="323" alt="image" src="https://github.com/OAI/OpenAPI-Specification/assets/85643503/c928faf6-3315-4fd7-a1a4-080e9f9f6015">

ralfhandl commented on versions/3.0.4.md:
```diff
@@ -17,74 +17,108 @@ For examples of OpenAPI usage and additional documentation, please visit [learn.
+    + [Response Object](#response-object)
+    + [Callback Object](#callback-object)
+    + [Example Object](#example-object)
+    + [Working With Examples](#working-with-examples)
```
Our respec build script actually removes this part of the Markdown, and maybe we should already do that in the source:

https://github.com/OAI/OpenAPI-Specification/blob/6bd37a356740b8128bfcca02b2d12d745def5135/scripts/md2html/md2html.js#L176-L178

handrews replied:
Does vanilla Respec do this?  We want to migrate to the regular tool (as we're using it on Moonwalk) eventually.

lornajane replied:
This is pretty interesting! I can just rip out the ToC as the markdown is no longer our authoritative source and on github it comes with ToC in the rendering

ralfhandl replied:
Vanilla ReSpec should work fine for us, see https://ralfhandl.github.io/respec-experiments/

ralfhandl replied:
Combined my experiment with Darrel's work in https://github.com/OAI/sig-moonwalk/pull/140

lornajane replied:
@ralfhandl nice!! However I'm deeply reluctant to introduce more powershell to the main OAI repository so we would need a (probably bash, looking at our current scripts and target contributor/participant audience) alternative.

ralfhandl replied:
@lornajane In the OpenAPI-Specification repository the ReSpec preparation is done via Bash and JavaScript, and I'd keep it that way.

Only replace the outdated and _modified_ ReSpec snapshot with a newer and _unmodified_ snapshot, or don't use a snapshot and rely on the current "live" version of ReSpec.

miqui reviewed (commented):
I have gone through the entire PR !  Formatting looks good.  Nice @lornajane !!

lornajane commented:
About the TOC, if I understood the thread correctly, respec builds a TOC as well, so we do not need one in our source at all

lornajane commented:
I'm not sure respec does the table of contents in a way we can use, so I've regenerated it in this update. Thanks @ralfhandl for getting into the respec details, I hadn't got there yet!

Changed items:
 - rebased to pick up newest additions to the dev branch
 - switched required bullet point format to `*` and make the toc command use this format as well
 - reapplied all other changes

It did occur to me that we could keep the `<!-- toc -->` notation in the source and just generate that bit when we're building the HTML version. It's the last commit on the branch here anyway, so super easy to remove.

ralfhandl commented:
> I'm not sure respec does the table of contents in a way we can use, so I've regenerated it in this update.

Our HTML build script removes the table of contents from the Markdown, the ToC we see for example in https://spec.openapis.org/oas/latest.html is generated by ReSpec from the section headlines.

Here's the relevant part of our build script:

https://github.com/OAI/OpenAPI-Specification/blob/6bd37a356740b8128bfcca02b2d12d745def5135/scripts/md2html/md2html.js#L176-L178

Every line between lines starting with `## Table of Contents` and `<!-- /TOC` is removed, including these lines.

With the current change from `<!-- /TOC` to `<!-- tocstop -->` everything except the first 18 lines would be removed unless we adjust the build script to the new ToC tool.

A better way forward would be to completely remove the "Table of Contents" subsection because
- we don't need it for producing the published HTML
- we don't need it for viewing the raw `*.md` files in GitHub because the GitHub Markdown viewer now has an Outline pane on the right with an auto-generated table of contents.

lornajane commented on versions/3.0.4.md:
```diff
@@ -17,80 +17,35 @@ For examples of OpenAPI usage and additional documentation, please visit [learn.
 For extension registries and other specifications published by the OpenAPI Initiative, as well as the authoritative rendering of this specification, please visit [spec.openapis.org](https://spec.openapis.org/).
 
 ## Table of Contents
```
Remove the heading

lornajane commented:
Should also tackle #1720

mikekistler reviewed (approved):
Looks good. 👍 

I reviewed this by performing all the automated changes myself on a fresh branch and then reviewing the (much smaller) diff.

Thank you @lornajane for this work.

mikekistler commented:
My notes on the process that @lornajane put in the PR description:

get checkout v3.0.4-dev
git checkout -b fix-markdown
cd versions

1. Run prettier for formatting

npx prettier --write --single-quote 3.0.4.md

2. Run [markdownlint](https://github.com/DavidAnson/markdownlint-cli2) to fix whatever it can

# Create markdown-lint.yaml with contents given, then
npx markdownlint-cli --config markdown-lint.yaml --fix 3.0.4.md

4. Take out the table of contents and its comments, replace with a single <!-- toc -->

awk '/## Table of Contents/{f=1} /<!-- \/TOC -->/{f=0; print "<!-- toc -->"; next} {if (f==0) {print}} ' 3.0.4.md > temp.md; mv temp.md 3.0.4.md

5. Run a magical one-off script to update/fix/rewrite/remove all our anchors and internal links.

python kebab_it.py 3.0.4.md > temp.md; mv temp.md 3.0.4.md
npx markdownlint-cli --fix 3.0.4.md

  - use markdownlint again to check that this all worked because it can check internal links

npx markdownlint-cli --config markdown-lint.yaml --fix 3.0.4.md

ralfhandl commented:
Can we use

> npx prettier --write --single-quote --prose-wrap preserve 3.0.4.md

to make diffing with the previous spec versions easier?

lornajane commented:
Updated to apply the changes to the newest 3.0.4 version, and marked as ready to review as I don't think we have any more changes in flight for this branch.

mikekistler reviewed (approved):
Renewing my approval. 👍

ralfhandl reviewed (approved):
Also checked the generated HTML:
* internal links use the new anchors
* all "code" blocks are prettier-formatted
* no further differences (after fixing the build script 😁)
